ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,363 @@
+introduced. This new type is located on `.spec.sources` slice, and triggers the same behavior.
+
+Additionally, the `timeout` attribute sets how long the `waiter` will hold back for user upload data.
+
+#### Waiting Mechanism
+
+The container-image built by Shipwright is created by a series of steps running sequentially, those
+steps are executed in the build-pod, running each container and inspecting for return-code and
+errors.
+
+The task at hand, waiting for the users' data transfer, will follow the same basic structure, which
+means a simple application meant to wait for a specific condition, and exit gracefully afterwards.
+Here we may keep checking if a given "lock" file exists, when removed the application exists. It also
+must exit with an error on timeout.
+
+Additionally, we can employ this application to make verifications on the streamed data, like for
+instance checksum, ownership (UID/GID) and more.
+
+The implementation written for research and development can be found at [otaviof/waiter][Waiter]
+repository, and it is started by running "`waiter start`". When running it will keep watching the
+file `/var/tmp/waiter.pid`, and when removed, it will exit gracefully. The following example shows it
+running in the container-manager. In the following example the entrypoint receives the argument
+"`start`", and the waiting begins.
+
+```sh
+$ docker run --name=waiter --interactive otaviof/waiter:latest start
+2021/06/01 08:25:24 Saving PID '1' at '/var/tmp/waiter.pid'
+2021/06/01 08:25:24 Attempts '600'...
+2021/06/01 08:25:25 Attempts '599'...
+2021/06/01 08:25:26 Attempts '598'...
+2021/06/01 08:25:27 Done!
+```
+
+When the data transfer is complete, the CLI will execute "`waiter done`", via the same "`kubectl
+exec`" mechanism.
+
+```sh
+$ docker exec waiter waiter done
+2021/06/01 08:25:56 Removing pid-file '/var/tmp/waiter.pid' (1)
+```
+
+This procedure works on the current version of [Tekton (v0.22.0)][TektonV0220] used in Shipwright
+Build-Controller.
+
+##### Waiter on Controller Repository
+
+The `waiter` application will live on the [Build Controller repository][ControllerCmd], side by side
+with other applications, and producing a stand alone container-image. The image must be based on
+UBI8, following the same standards than others.
+
+This image also has `tar` installed, which will be used for the data upload mechanism, as explained
+before.
+
+Another requirement is following the same configuration standard than other external images used by
+the Build Controller, like a environment variable (i.e.: `GIT_CONTAINER_TEMPLATE`) to overwrite the
+image before the controller starts.
+
+At last, it should provide meaningful information for Tekton's `TaskResult`, so we can differentiate
+this step run from others. For the `waiter` process results, it can take advantage of the
+[termination-log pattern][KubernetesTerminationLog] to inform Tekton more detailed execution results.
+
+### Use-Cases
+
+#### Developer's Inner-Loop
```

Comment:
per comments above when you mentioned this one, this is one of those user stories ... not too surprising given you "use-cases" section title above :-)

Reply:
I'll make sure both places are consolidated, and the names match the template as well.

Reply:
I mean, the section has been renamed to "User Stories", and it's inside the "Proposal" section already. I was not planing to move it all the way up to the top, tbh. Do you think we should discuss the user-stores before discussing the proposal itself?

Reply:
moving this section is what the current SHIP template calls for

If none of the other reviewers object, and you feel strongly about it, I'll consent to not blocking the merge of this PR on that point.